File: src/main/java/io/lcydev/spawnercleaner/SpawnerCleaner.java
package io.lcydev.spawnercleaner;

import io.lcydev.spawnercleaner.utils.Config;
import io.lcydev.spawnercleaner.listeners.SpawnerListener;
import io.lcydev.spawnercleaner.listeners.PreSpawnerListener;
import io.papermc.lib.PaperLib;
import org.bukkit.Bukkit;
import org.bukkit.Location;
import org.bukkit.plugin.PluginManager;
import org.bukkit.plugin.java.JavaPlugin;
import org.bukkit.scheduler.BukkitScheduler;

import java.util.LinkedList;
import java.util.Queue;

public final class SpawnerCleaner extends JavaPlugin {

    //Instance Object to use when no Object was passed!
    private static SpawnerCleaner instance;
    // Queue blocks for removal in a modifiable List
    public Queue<Location> removeQueue = new LinkedList<>();
    // Does the server implement Paper
    private boolean isPaper;

    public SpawnerCleaner() {
        super();
        instance = this;
    }

    @Override
    public void onEnable() {
        isPaper = PaperLib.isPaper();
        PaperLib.suggestPaper(this);
        if (!Config.init()) {
            instance.getPluginLoader().disablePlugin(this);
        } else {
            Initialize();
            getLogger().info("SpawnerCleaner is now Enabled!");
        }
    }

    @Override
    public void onDisable() {
        getLogger().info("SpawnerCleaner is now Disabled!");
    }

    private void Initialize() {
        Config config = Config.get();
        BukkitScheduler scheduler = Bukkit.getScheduler();
        PluginManager pluginManager = Bukkit.getPluginManager();

        if (config.getMode().equalsIgnoreCase("EVENT")) {
            if (isPaper) {
                pluginManager.registerEvents(new PreSpawnerListener(), this);
            } else {
                pluginManager.registerEvents(new SpawnerListener(), this);
            }
        } else if (config.getMode().equalsIgnoreCase("CHUNK")) {
        }
        scheduler.scheduleSyncRepeatingTask(this, new RemoveTask(), 0L, 8L);
    }

    /**
     * Gets the instance of the SpawnerCleaner plugin.
     *
     * @return The instance of this plugin.
     */
    public static SpawnerCleaner inst() {
        return instance;
    }
}

File: src/main/kotlin/io/lcydev/spawnercleaner/SpawnerCleaner.kt
package io.lcydev.template;

import gg.flyte.twilight.twilight
import io.lcydev.template.command.PingCommand
import io.lcydev.template.handlers.ChunkManager
import io.lcydev.template.listeners.PreSpawnerListener
import io.lcydev.template.listeners.SpawnerListener
import org.bukkit.Location
import org.bukkit.plugin.PluginManager
import org.bukkit.plugin.java.JavaPlugin
import revxrsal.commands.bukkit.BukkitCommandHandler
import java.util.*
import kotlin.system.measureTimeMillis


class SpawnerCleaner : JavaPlugin() {

    companion object {
        const val CONFIG_VERSION = "1.1.0"

        lateinit var instance: SpawnerCleaner
            private set
    }

    private val pluginManager: PluginManager = server.pluginManager
    val chunkManager: ChunkManager = ChunkManager()

    // Queue blocks for removal in a modifiable List
    var removeQueue: Queue<Location> = LinkedList()

    override fun onLoad() {
        instance = this
    }

    override fun onEnable() {
        val timeTaken = measureTimeMillis {
            registerListeners()
        }
        logger.info("Plugin enabled in time $timeTaken ms")

        twilight(this) { }

        BukkitCommandHandler.create(this).apply {
            enableAdventure()
            register(PingCommand())
            registerBrigadier()
        }
    }

    private fun registerListeners() {
        val listeners = arrayListOf(
            SpawnerListener(this),
            PreSpawnerListener(this),
        )
        val timeTaken = measureTimeMillis {
            listeners.forEach { listener ->
                pluginManager.registerEvents(listener, this)
                logger.info("Bukkit Listener ${listener::class.simpleName} registered () -> ok")
            }
        }
        logger.info("Listeners registered (${listeners.size}) in time $timeTaken ms -> ok")
    }
}